File: microservers/billetterieservice/src/main/java/io/multi/billetterieservice/service/VilleService.java
package io.multi.billetterieservice.service;

import io.multi.billetterieservice.dto.VilleCreateRequest;
import io.multi.billetterieservice.dto.VilleUpdateRequest;
import io.multi.billetterieservice.response.VilleResponse;

import java.util.List;

/**
 * Service pour la gestion des villes.
 */
public interface VilleService {

    /**
     * Crée une nouvelle ville.
     *
     * @param request Les données de la ville à créer
     * @return La ville créée
     */
    VilleResponse createVille(VilleCreateRequest request);

    /**
     * Met à jour une ville.
     *
     * @param villeUuid L'UUID de la ville à mettre à jour
     * @param request   Les nouvelles données
     * @return La ville mise à jour
     */
    VilleResponse updateVille(String villeUuid, VilleUpdateRequest request);

    /**
     * Active ou désactive une ville.
     *
     * @param villeUuid L'UUID de la ville
     * @param actif     Le nouveau statut
     * @return La ville mise à jour
     */
    VilleResponse updateVilleStatus(String villeUuid, Boolean actif);

    /**
     * Récupère toutes les villes.
     *
     * @return Liste de toutes les villes
     */
    List<VilleResponse> getAllVilles();

    /**
     * Récupère toutes les villes actives.
     *
     * @return Liste des villes actives
     */
    List<VilleResponse> getActiveVilles();

    /**
     * Récupère les villes d'une région.
     *
     * @param regionUuid L'UUID de la région
     * @return Liste des villes de la région
     */
    List<VilleResponse> getVillesByRegion(String regionUuid);

    /**
     * Récupère les villes actives d'une région.
     *
     * @param regionUuid L'UUID de la région
     * @return Liste des villes actives de la région
     */
    List<VilleResponse> getActiveVillesByRegion(String regionUuid);

    /**
     * Récupère une ville par son UUID.
     *
     * @param villeUuid L'UUID de la ville
     * @return La ville trouvée
     */
    VilleResponse getVilleByUuid(String villeUuid);
}

File: frontend/ultima-ng-19.0.0/src/app/service/ville.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { catchError, Observable, tap, throwError } from 'rxjs';
import { server } from '@/utils/fileutils';
import { IResponse } from '@/interface/response';
import { IVilleCreateRequest, IVilleUpdateRequest } from '@/interface/ville';

@Injectable({
    providedIn: 'root'
})
export class VilleService {
    private http = inject(HttpClient);
    private baseUrl = `${server}/billetterie/villes`;

    constructor() {}

    /**
     * Récupère toutes les villes
     */
    getAllVilles$ = (): Observable<IResponse> => this.http.get<IResponse>(this.baseUrl).pipe(tap(console.log), catchError(this.handleError));

    /**
     * Récupère uniquement les villes actives
     */
    getActiveVilles$ = (): Observable<IResponse> => this.http.get<IResponse>(`${this.baseUrl}/active`).pipe(tap(console.log), catchError(this.handleError));

    /**
     * Récupère une ville par son UUID
     */
    getVilleByUuid$ = (villeUuid: string): Observable<IResponse> => this.http.get<IResponse>(`${this.baseUrl}/${villeUuid}`).pipe(tap(console.log), catchError(this.handleError));

    /**
     * Récupère les villes d'une région
     */
    getVillesByRegion$ = (regionUuid: string): Observable<IResponse> => this.http.get<IResponse>(`${this.baseUrl}/region/${regionUuid}`).pipe(tap(console.log), catchError(this.handleError));

    /**
     * Récupère les villes actives d'une région
     */
    getActiveVillesByRegion$ = (regionUuid: string): Observable<IResponse> => this.http.get<IResponse>(`${this.baseUrl}/region/${regionUuid}/active`).pipe(tap(console.log), catchError(this.handleError));

    /**
     * Crée une nouvelle ville
     */
    createVille$ = (request: IVilleCreateRequest): Observable<IResponse> => this.http.post<IResponse>(this.baseUrl, request).pipe(tap(console.log), catchError(this.handleError));

    /**
     * Met à jour une ville
     */
    updateVille$ = (villeUuid: string, request: IVilleUpdateRequest): Observable<IResponse> => this.http.put<IResponse>(`${this.baseUrl}/${villeUuid}`, request).pipe(tap(console.log), catchError(this.handleError));

    /**
     * Active ou désactive une ville
     */
    updateVilleStatus$ = (villeUuid: string, actif: boolean): Observable<IResponse> => this.http.patch<IResponse>(`${this.baseUrl}/${villeUuid}/status`, { actif }).pipe(tap(console.log), catchError(this.handleError));

    /**
     * Gestion des erreurs HTTP
     */
    private handleError = (httpErrorResponse: HttpErrorResponse): Observable<never> => {
        console.error('VilleService Error:', httpErrorResponse);
        let error: string = 'Une erreur est survenue. Veuillez réessayer.';

        if (httpErrorResponse.error instanceof ErrorEvent) {
            error = `Erreur client - ${httpErrorResponse.error.message}`;
            return throwError(() => error);
        }

        if (httpErrorResponse.error?.message) {
            error = httpErrorResponse.error.message;
            return throwError(() => error);
        }

        if (httpErrorResponse.status === 401) {
            error = 'Session expirée. Veuillez vous reconnecter.';
        } else if (httpErrorResponse.status === 403) {
            error = 'Accès non autorisé.';
        } else if (httpErrorResponse.status === 404) {
            error = 'Ville non trouvée.';
        } else if (httpErrorResponse.status === 409) {
            error = 'Cette ville existe déjà dans cette région.';
        }

        return throwError(() => error);
    };
}
